crates/lib3h/src/rrdht_util/arc.rs

```diff
@@ -0,0 +1,308 @@
+use crate::rrdht_util::Location;
+
+pub const ARC_LENGTH_MAX: u64 = 0x100000000;
+pub const ARC_RADIUS_MAX: u32 = 0x80000001;
+
+/// An rrdht "arc" indicates a range on the u32 "location" circle
+#[derive(Clone, Copy, PartialEq, Eq)]
+pub struct Arc {
+    start: Location,
+    // length must be a u64 because we need to be able to represent
+    // both a zero length arc with length 0,
+    // and also a full arc with length one beyond the u32 max
+    length: u64,
+}
+
+impl std::fmt::Display for Arc {
+    fn fmt(&self, f: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
+        write!(f, "32r{:08x}:{:09x}", u32::from(self.start), self.length)
+    }
+}
+
+impl std::fmt::Debug for Arc {
+    fn fmt(&self, f: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
+        f.debug_tuple("Arc").field(&format!("{}", self)).finish()
+    }
+}
+
+impl Arc {
+    /// construct a new Arc instance given a start Location and a length
+    pub fn new(start: Location, length: u64) -> Self {
+        let length = std::cmp::min(length, ARC_LENGTH_MAX);
+        Self { start, length }
+    }
+
+    /// construct a new Arc instance given a center Location and a radius
+    pub fn new_radius(center: Location, radius: u32) -> Self {
+        if radius == 0 {
+            return Arc::new(center, 0);
+        }
+        let radius = std::cmp::min(radius, ARC_RADIUS_MAX);
+        let start = center - Location::from(radius - 1);
+        let length = u64::from(radius) * 2 - 1;
+        Arc::new(start, length)
+    }
+
+    /// construct a new Arc from canonical BITSrHEX:HEX format
+    pub fn new_repr(repr: &str) -> Self {
+        let pre = &repr[0..3];
+        assert_eq!("32r", pre);
+        let loc = u32::from_str_radix(&repr[3..11], 16).expect("can parse hex");
+        let len = u64::from_str_radix(&repr[12..], 16).expect("can parse hex");
+        Self {
+            start: loc.into(),
+            length: len,
+        }
+    }
+
+    /// the start position for this arc
+    pub fn start(&self) -> Location {
+        self.start
+    }
+
+    /// the length of this arc
+    pub fn length(&self) -> u64 {
+        self.length
+    }
+
+    /// the center point of this arc, the "location" if specified by radius
+    pub fn center(&self) -> Location {
+        if self.length == 0 {
+            return self.start;
+        }
+        self.start + (self.radius() - 1).into()
+    }
+
+    /// the radius of this arc
+    pub fn radius(&self) -> u32 {
+        if self.length == 0 {
+            return 0;
+        }
+        (self.length / 2 + 1) as u32
```

Comment:
Is the +1 to account for the remainder when lengths are odd? If so this seems wrong for even lengths. But maybe this is an offset instead. Regardless a comment might be helpful here.

Reply:
This is actually to reverse the `- 1` when [creating an arc](https://github.com/holochain/lib3h/blob/rrdht-util/crates/lib3h/src/rrdht_util/arc.rs#L42) based off a center point and a "radius" (erm... half the length?). We this value to mean not storing anything when zero and only store the exact center point when 1, which requires this subtraction / addition.

Will try to write a less ramble-y comment : )